Finding best model and hyper parameter tunning using GridSearchCV and RandomizedSearchCV

In [27]:
import numpy as np 
import pandas as pd
import matplotlib.pyplot as plt
%matplotlib inline
from sklearn.datasets import load_digits
digits=load_digits()
dir(digits)
df=pd.DataFrame(digits['data'],columns=digits['feature_names'])
df

pixel_0_0  pixel_0_1  pixel_0_2  pixel_0_3  pixel_0_4  pixel_0_5  \
0           0.0        0.0        5.0       13.0        9.0        1.0   
1           0.0        0.0        0.0       12.0       13.0        5.0   
2           0.0        0.0        0.0        4.0       15.0       12.0   
3           0.0        0.0        7.0       15.0       13.0        1.0   
4           0.0        0.0        0.0        1.0       11.0        0.0   
...         ...        ...        ...        ...        ...        ...   
1792        0.0        0.0        4.0       10.0       13.0        6.0   
1793        0.0        0.0        6.0       16.0       13.0       11.0   
1794        0.0        0.0        1.0       11.0       15.0        1.0   
1795        0.0        0.0        2.0       10.0        7.0        0.0   
1796        0.0        0.0       10.0       14.0        8.0        1.0   

      pixel_0_6  pixel_0_7  pixel_1_0  pixel_1_1  ...  pixel_6_6  pixel_6_7  \
0           0.0        0.0        0.0        0.0  ...        0.0        0.0   
1           0.0        0.0        0.0        0.0  ...        0.0        0.0   
2           0.0        0.0        0.0        0.0  ...        5.0        0.0   
3           0.0        0.0        0.0        8.0  ...        9.0        0.0   
4           0.0        0.0        0.0        0.0  ...        0.0        0.0   
...         ...        ...        ...        ...  ...        ...        ...   
1792        0.0        0.0        0.0        1.0  ...        4.0        0.0   
1793        1.0        0.0        0.0        0.0  ...        1.0        0.0   
1794        0.0        0.0        0.0        0.0  ...        0.0        0.0   
1795        0.0        0.0        0.0        0.0  ...        2.0        0.0   
1796        0.0        0.0        0.0        2.0  ...        8.0        0.0   

      pixel_7_0  pixel_7_1  pixel_7_2  pixel_7_3  pixel_7_4  pixel_7_5  \
0           0.0        0.0        6.0       13.0       10.0        0.0   
1           0.0        0.0        0.0       11.0       16.0       10.0   
2           0.0        0.0        0.0        3.0       11.0       16.0   
3           0.0        0.0        7.0       13.0       13.0        9.0   
4           0.0        0.0        0.0        2.0       16.0        4.0   
...         ...        ...        ...        ...        ...        ...   
1792        0.0        0.0        2.0       14.0       15.0        9.0   
1793        0.0        0.0        6.0       16.0       14.0        6.0   
1794        0.0        0.0        2.0        9.0       13.0        6.0   
1795        0.0        0.0        5.0       12.0       16.0       12.0   
1796        0.0        1.0        8.0       12.0       14.0       12.0   

      pixel_7_6  pixel_7_7  
0           0.0        0.0  
1           0.0        0.0  
2           9.0        0.0  
3           0.0        0.0  
4           0.0        0.0  
...         ...        ...  
1792        0.0        0.0  
1793        0.0        0.0  
1794        0.0        0.0  
1795        0.0        0.0  
1796        1.0        0.0  

[1797 rows x 64 columns]

Importing All models which will be checked

In [28]:
from sklearn.linear_model import LogisticRegression
from sklearn.ensemble import RandomForestClassifier
from sklearn.naive_bayes import GaussianNB
from sklearn.naive_bayes import MultinomialNB
from sklearn import tree
from sklearn import svm

GridSearchCV

In [29]:
from sklearn.model_selection import GridSearchCV
clf = GridSearchCV(
    LogisticRegression(),
    {
        'C': [1, 10, 100],
        'tol': [0.0001],
        'max_iter': [100, 200]
    },
    cv=5,
    return_train_score=False
)
clf.fit(digits['data'],digits['target'])
clf.cv_results_

c:\Users\HP USER\Anaconda3\Lib\site-packages\sklearn\linear_model\_logistic.py:469: ConvergenceWarning: lbfgs failed to converge (status=1):
STOP: TOTAL NO. of ITERATIONS REACHED LIMIT.

Increase the number of iterations (max_iter) or scale the data as shown in:
    https://scikit-learn.org/stable/modules/preprocessing.html
Please also refer to the documentation for alternative solver options:
    https://scikit-learn.org/stable/modules/linear_model.html#logistic-regression
  n_iter_i = _check_optimize_result(
c:\Users\HP USER\Anaconda3\Lib\site-packages\sklearn\linear_model\_logistic.py:469: ConvergenceWarning: lbfgs failed to converge (status=1):
STOP: TOTAL NO. of ITERATIONS REACHED LIMIT.

Increase the number of iterations (max_iter) or scale the data as shown in:
    https://scikit-learn.org/stable/modules/preprocessing.html
Please also refer to the documentation for alternative solver options:
    https://scikit-learn.org/stable/modules/linear_model.html#logistic-regression
  n_i

{'mean_fit_time': array([0.03960233, 0.06464267, 0.04300323, 0.04807692, 0.0256289 ,
        0.02376094]),
 'std_fit_time': array([0.00154685, 0.00879697, 0.0015568 , 0.00650088, 0.00406213,
        0.00329803]),
 'mean_score_time': array([0.00066905, 0.00080023, 0.00083094, 0.00081697, 0.00060296,
        0.00082936]),
 'std_score_time': array([0.00056062, 0.00040011, 0.00042466, 0.00040979, 0.00049236,
        0.00041846]),
 'param_C': masked_array(data=[1, 1, 10, 10, 100, 100],
              mask=[False, False, False, False, False, False],
        fill_value=999999),
 'param_max_iter': masked_array(data=[100, 200, 100, 200, 100, 200],
              mask=[False, False, False, False, False, False],
        fill_value=999999),
 'param_tol': masked_array(data=[0.0001, 0.0001, 0.0001, 0.0001, 0.0001, 0.0001],
              mask=[False, False, False, False, False, False],
        fill_value=1e+20),
 'params': [{'C': 1, 'max_iter': 100, 'tol': 0.0001},
  {'C': 1, 'max_iter': 200, 'tol': 0.

Using all combinations total (4*3*4)=48 differnet combinations are generated

In [30]:
df1=pd.DataFrame(clf.cv_results_)
df1


mean_fit_time  std_fit_time  mean_score_time  std_score_time  param_C  \
0       0.039602      0.001547         0.000669        0.000561        1   
1       0.064643      0.008797         0.000800        0.000400        1   
2       0.043003      0.001557         0.000831        0.000425       10   
3       0.048077      0.006501         0.000817        0.000410       10   
4       0.025629      0.004062         0.000603        0.000492      100   
5       0.023761      0.003298         0.000829        0.000418      100   

   param_max_iter  param_tol                                      params  \
0             100     0.0001    {'C': 1, 'max_iter': 100, 'tol': 0.0001}   
1             200     0.0001    {'C': 1, 'max_iter': 200, 'tol': 0.0001}   
2             100     0.0001   {'C': 10, 'max_iter': 100, 'tol': 0.0001}   
3             200     0.0001   {'C': 10, 'max_iter': 200, 'tol': 0.0001}   
4             100     0.0001  {'C': 100, 'max_iter': 100, 'tol': 0.0001}   
5             200     0.0001  {'C': 100, 'max_iter': 200, 'tol': 0.0001}   

   split0_test_score  split1_test_score  split2_test_score  split3_test_score  \
0           0.922222           0.869444           0.941504           0.938719   
1           0.922222           0.872222           0.941504           0.941504   
2           0.916667           0.858333           0.941504           0.941504   
3           0.919444           0.858333           0.935933           0.933148   
4           0.919444           0.872222           0.941504           0.938719   
5           0.919444           0.872222           0.941504           0.938719   

   split4_test_score  mean_test_score  std_test_score  rank_test_score  
0           0.896936         0.913765        0.027244                4  
1           0.896936         0.914878        0.026882                3  
2           0.894150         0.910432        0.031462                5  
3           0.894150         0.908202        0.028990                6  
4           0.910864         0.916551        0.024979                1  
5           0.910864         0.916551        0.024979                1

In [31]:
df1[['param_C','param_tol','param_max_iter','mean_test_score']]


param_C  param_tol  param_max_iter  mean_test_score
0        1     0.0001             100         0.913765
1        1     0.0001             200         0.914878
2       10     0.0001             100         0.910432
3       10     0.0001             200         0.908202
4      100     0.0001             100         0.916551
5      100     0.0001             200         0.916551

In [32]:
clf.best_score_

0.9165506035283194

In [33]:
clf.best_params_

{'C': 100, 'max_iter': 100, 'tol': 0.0001}

Use RandomizedSearchCV to reduce number of iterations and with random combination of parameters. This is useful when you have too many parameters to try and your training time is longer. It helps reduce the cost of computation

In [34]:
from sklearn.model_selection import RandomizedSearchCV
clf2 = RandomizedSearchCV(
    LogisticRegression(),
    {
        'C': [1, 10, 100, 1000],
        'tol': [0.0001, 0.00001, 0.000001],
        'max_iter': [100, 200, 300, 400]
    },
    cv=5,
    return_train_score=False,
    n_iter=10
)
clf2.fit(digits['data'],digits['target'])


c:\Users\HP USER\Anaconda3\Lib\site-packages\sklearn\linear_model\_logistic.py:469: ConvergenceWarning: lbfgs failed to converge (status=1):
STOP: TOTAL NO. of ITERATIONS REACHED LIMIT.

Increase the number of iterations (max_iter) or scale the data as shown in:
    https://scikit-learn.org/stable/modules/preprocessing.html
Please also refer to the documentation for alternative solver options:
    https://scikit-learn.org/stable/modules/linear_model.html#logistic-regression
  n_iter_i = _check_optimize_result(
c:\Users\HP USER\Anaconda3\Lib\site-packages\sklearn\linear_model\_logistic.py:469: ConvergenceWarning: lbfgs failed to converge (status=1):
STOP: TOTAL NO. of ITERATIONS REACHED LIMIT.

Increase the number of iterations (max_iter) or scale the data as shown in:
    https://scikit-learn.org/stable/modules/preprocessing.html
Please also refer to the documentation for alternative solver options:
    https://scikit-learn.org/stable/modules/linear_model.html#logistic-regression
  n_i

RandomizedSearchCV(cv=5, estimator=LogisticRegression(),
                   param_distributions={'C': [1, 10, 100, 1000],
                                        'max_iter': [100, 200, 300, 400],
                                        'tol': [0.0001, 1e-05, 1e-06]})

In [35]:
pd.DataFrame(clf2.cv_results_)[['param_C','param_tol','param_max_iter','mean_test_score']]

param_C  param_tol  param_max_iter  mean_test_score
0      100   0.000010             100         0.911541
1       10   0.000001             400         0.908197
2       10   0.000100             100         0.910432
3      100   0.000010             400         0.907645
4      100   0.000001             200         0.908756
5        1   0.000010             200         0.914322
6     1000   0.000001             400         0.907088
7        1   0.000010             300         0.912094
8       10   0.000001             100         0.910987
9        1   0.000100             400         0.914878

In [36]:
clf2.best_score_

0.9148777468276075

In [37]:
clf2.best_params_

{'tol': 0.0001, 'max_iter': 400, 'C': 1}

How about different models with different hyperparameters?

In [38]:
model_params = {
    'svm': {
        'model': svm.SVC(gamma='auto'),
        'params': {
            'C': [1, 10, 20],
            'kernel': ['rbf', 'linear']
        }
    },
    'random_forest': {
        'model': RandomForestClassifier(),
        'params': {
            'n_estimators': [1, 5, 10]
        }
    },
    'logistic_regression': {
        'model': LogisticRegression(solver='liblinear', multi_class='auto'),
        'params': {
            'C': [1, 5, 10]
        }
    },
    'decision_tree': {
        'model': tree.DecisionTreeClassifier(),
        'params': {
            'criterion': ['gini', 'entropy', 'log_loss']
        }
    },
    'gaussian_nb': {
        'model': GaussianNB(),
        'params': {}
    },
    'multinomial_nb': {
        'model': MultinomialNB(),
        'params': {}
    }
}

In [39]:
scores=[]

for model_name, mp in model_params.items():
    clf_new =  GridSearchCV(mp['model'], mp['params'], cv=5, return_train_score=False)
    clf_new.fit(digits['data'], digits['target'])
    scores.append({
        'model': model_name,
        'best_score': clf_new.best_score_,
        'best_params': clf_new.best_params_
    })
    
df_new= pd.DataFrame(scores,columns=['model','best_score','best_params'])
df_new

c:\Users\HP USER\Anaconda3\Lib\site-packages\sklearn\linear_model\_logistic.py:1247: FutureWarning: 'multi_class' was deprecated in version 1.5 and will be removed in 1.7. From then on, it will always use 'multinomial'. Leave it to its default value to avoid this warning.
  warnings.warn(
c:\Users\HP USER\Anaconda3\Lib\site-packages\sklearn\linear_model\_logistic.py:1247: FutureWarning: 'multi_class' was deprecated in version 1.5 and will be removed in 1.7. From then on, it will always use 'multinomial'. Leave it to its default value to avoid this warning.
  warnings.warn(
c:\Users\HP USER\Anaconda3\Lib\site-packages\sklearn\linear_model\_logistic.py:1247: FutureWarning: 'multi_class' was deprecated in version 1.5 and will be removed in 1.7. From then on, it will always use 'multinomial'. Leave it to its default value to avoid this warning.
  warnings.warn(
c:\Users\HP USER\Anaconda3\Lib\site-packages\sklearn\linear_model\_logistic.py:1247: FutureWarning: 'multi_class' was deprecated i

model  best_score                   best_params
0                  svm    0.947697  {'C': 1, 'kernel': 'linear'}
1        random_forest    0.907640          {'n_estimators': 10}
2  logistic_regression    0.922114                      {'C': 1}
3        decision_tree    0.813049      {'criterion': 'entropy'}
4          gaussian_nb    0.806928                            {}
5       multinomial_nb    0.870350                            {}

Using Hyper Parameter tuning it is found that the best model is "svm" with best_params {'C': 1, 'kernel': 'linear'} and best_score is 0.947697